requests-for-comments/2024-01-23-project-environments.md

```diff
@@ -0,0 +1,116 @@
+
+We currently solve this in Feature flags with a feature for syncing flags between projects, based on the flag key. Feature flags make this uniquely easy thanks to the feature flag key being an effective organization-wide identifier. No other entity – dashboards, insights, etc. – has this advantage though.
+
+Because feature flags use cohorts and actions in their definitions, these are also synced in the current solution –but only based on the cohort/action name, so if the name is changed in one environment, the link is silently broken.
+
+
+Previous context on this topic: [the December 2022 "Feature flag environments" RFC](https://github.com/PostHog/product-internal/blob/main/requests-for-comments/2022-12-08-feature-flag-environments.md#221115---rfc-feature-flag-environments)
+
+## Expectations
+
+### Isolated
+
+The key feature is separation of all of *data*:
+
+- Events
+- People
+- Groups
+- Recordings
+
+Pipeline setup should probably be separate too:
+
+- Apps
+- Batch exports
+
+Domain-specific configuration is also going to differ between environments likely:
+
+- Toolbar authorized URLs
+
+### Shared
+
+Sharing of these entities is commonly requested:
+
+- Dashboards
+- Insights
+- Actions
+- Cohorts
+- Feature flags
+
+Not mentioned typically, but logically also shared should be:
+
+- Event and property definitions
+- Group types
+- Annotations
+- Playlists
+- Experiments
+- Surveys
+- Early access features
+- Project settings
+
+The needs are unclear for:
+
+- Data warehouse tables
+
+## Rejected solution: Cross-project copying
+
+This is what some users have asked for verbatim. What it would probably mean is a feature flags-style "Projects" view on the item level, allowing an item to be copied/synced across projects. The reliability of syncing is key though, and the problem mentioned before is that there's no inherent cross-project ID for this. Just basing on the entity name would be extremely fragile – there needs to be an underlying link better than that.
+
+This means we'd need to maintain such underlying links (e.g. a cross-project ID, adjacent to the primary key ID) for all entities we'd like to work cross-project. In one way, this could be nice if you'd like to develop, say, a dashboard in the test environment and only copy it to the production environment when final. Divergence becomes easy. But there would be an act of manual syncing, making the experience cumbersome for users at scale. This would work, but it wouldn't be turnkey.
+
+[insert metric for how often flags differ within org]
+
+## Proposed solution: Environments
+
+I believe to solve this problem once and for all, we need environments to be a first-class feature, specifically: subdivisions of a project.
```

Comment:
Would love to hear more on this, and what it looks like for the end user. Does it mean a separate copy of Cohort with ID 1234 per environment? And users can edit it per environment? Do they still have to 'sync' between environments? Or is there no copy, and the same cohort always takes the same values in all environments?

Reply:
There is no copy – the cohort is defined once in the project. It's like a universal version of the "Filter out internal and test users" toggle, except you define what's "internal and test" when initializing the client library, and not at query time.

Reply:
Hmm, I see, this is a bit scary for any products that affect the end-users app, like flags and surveys: Making changes in the dev environment _shouldn't_ affect production at all.

At the same time, I can see how this can be very nice for dashboards - can't imagine wanting different dashboards in different environments. Although, even here, I think it will be ~~scary~~ annoying if I'm debugging some event issue in dev and end up saving the dashboard, which then causes the production dashboard to update too.

I'm not sure this works well when environments are different country-based envs for example. In here you'll necessarily have different filters on the same dashboard, depending on the environment?

Reply:
I reviewed more tickets and chatted with some customers. I included some of this research, but it looks like they mostly want dashboards/insights/cohorts/etc. to be environment-specific, with select ones being promoted to production. I think that'll be a safe and intuitive workflow where we still avoid all the woes of copying – included in the RFC